### ISSUE TITLE
Discussion on Optimizer class

### ISSUE THREAD

**Issue by andyfaff** [andyfaff/scipy#4]

**Comment by stsievert** [comment]
That was quick!

Do we want `OptimizeFunction` to recursively call itself? I tried this but was unsure which method was "better". But seeing the implementation, maybe this could clean up some of the stuff around the number of evaluations (through `nfev`, `nhev`, `njev`).

``` python
class Function:
    def __init__(self, fn, grad=None, hess=None):
        self.func = func
        if grad is not None:
            self.grad = Function(grad, grad=hess)
        if hess is not None:
            self.hess = Function(hess)
        def __call__(self, *args, **kwargs):
            self.evals += 1
            self._func(*args, **kwargs)
        def _func(self, *args, **kwargs):
            self.func(*((x,) + self.args))
```

**Comment by andyfaff** [comment]
My thoughts were that for a given system `func, grad, hess` should be accessible from the same object. If `jac` and `hess` functions aren't supplied then they should be calculated numerically from that object.

If the user inherits OptimizeFunction then they can override jac and hess`, in which case they are ignored.

**Comment by andyfaff** [comment]
I tried a variety of things to try and keep the counting for fev, jev, hev, but it's a bit tricky when you want to keep subclassing, and also retain the option of supplying individual functions

**Comment by stsievert** [comment]
> want to keep subclassing

Yeah, I went through the same thoughts. I think easy subclassing is more important than the small benefit it'd bring us.
